### PR TITLE
[FLINK-29736] Abstract a table interface for both data and metadata tables

### PR DESCRIPTION
**Pull request by JingsongLi**
Currently, FileStoreTable is only for data tables, we can create a new interface Table for all tables.

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java [hunk 1]
```diff
@@ -116,7 +117,7 @@ void dropDatabase(String name, boolean ignoreIfNotExists, boolean cascade)
      * @return The requested table
      * @throws TableNotExistException if the target does not exist
      */
-    default FileStoreTable getTable(ObjectPath tablePath) throws TableNotExistException {
+    default Table getTable(ObjectPath tablePath) throws TableNotExistException {
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/catalog/Catalog.java:120]: Better to update the method description at L#114
```
    Return a {@link Table} identified by ...
```

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/DataTableScan.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        scan.withSnapshot(snapshotId);
+        return this;
+    }
+
+    public DataTableScan withFilter(List<Predicate> predicates) {
+        if (predicates == null || predicates.isEmpty()) {
+            return this;
+        }
+        return withFilter(PredicateBuilder.and(predicates));
+    }
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/DataTableScan.java:65]: Is there any concern not to use super interface's `withFilter(List<Predicate> predicates)`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/DataTableScan.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    public DataTableScan withFilter(Predicate predicate) {
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/DataTableScan.java:67]: Add `@Override` annotation?

File: flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkTable.java [hunk 4]
```diff
@@ -33,12 +32,12 @@
 import java.util.HashSet;
 import java.util.Set;
 
-/** A spark {@link Table} for table store. */
+/** A spark {@code Table} for table store. */
```
Comment (LadyForest) [inline flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkTable.java:35]: Why change it?

File: flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkScan.java [hunk 5]
```diff
@@ -128,13 +124,13 @@ public boolean equals(Object o) {
         }
 
         SparkScan that = (SparkScan) o;
-        return table.location().equals(that.table.location())
+        return table.name().equals(that.table.name())
```
Comment (LadyForest) [inline flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkScan.java:127]: Is it safe to just compare table's name? What about two tables with the same name and schema but under different namespace?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/DataTableScan.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/DataTableScan.java:65]: No, let's use it.

File: flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkTable.java (hunk 4 above)
Comment (JingsongLi) [reply flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkTable.java:35]: I will change it to `org.apache.spark.sql.connector.catalog.Table`

File: flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkScan.java (hunk 5 above)
Comment (JingsongLi) [reply flink-table-store-spark/src/main/java/org/apache/flink/table/store/spark/SparkScan.java:127]: In one catalog, different tables should have different names.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/TableStreamingReader.java [hunk 6]
```diff
@@ -84,11 +84,11 @@ public TableStreamingReader(
     @Nullable
     public Iterator<RowData> nextBatch() throws IOException {
         if (enumerator == null) {
-            TableScan scan = table.newScan();
+            DataTableScan scan = (DataTableScan) table.newScan();
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/TableStreamingReader.java:87]: Remove redundant cast?

**Review by LadyForest — APPROVED** [review]
Thanks for the update, looks good to me :D

**Pull request merged**